packages/core/src/execute/BaseRoutine.ts

```diff
@@ -38,19 +38,29 @@ export default abstract class BaseRoutine<Ctx extends Context<BaseContextArgs>>
 
   async execute(context: Ctx): Promise<any[]> {
     const value = await this.serializeTasks();
-    const { other, priority } = this.orderByWorkspacePriorityGraph();
-
-    await this.serializeRoutines(value, priority);
+    const batches = this.orderByWorkspacePriorityGraph();
+    const errors: Error[] = [];
 
     const concurrency = context.args.concurrency || this.tool.config.execute.concurrency;
-    const response = await this.poolRoutines(value, concurrency ? { concurrency } : {}, other);
+    const responses: AggregatedResponse[] = [];
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const batch of batches) {
+      // eslint-disable-next-line no-await-in-loop
+      const response = await this.poolRoutines(value, concurrency ? { concurrency } : {}, batch);
+
+      responses.push(response);
+
+      if (response.errors.length > 0) {
+        errors.push(...response.errors);
+      }
+    }
 
-    if (response.errors.length > 0) {
-      this.formatAndThrowErrors(response.errors);
+    if (errors.length > 0) {
+      this.formatAndThrowErrors(errors);
     }
 
-    // Not running in workspaces, so return value directly
-    return context.args.workspaces ? response.results : response.results[0];
+    return responses[0].results[0];
```

Comment:
We should return all results when using `--workspaces`, otherwise just the first result.

Reply:
updated